File: android/Showcase/src/main/java/com/spruceid/mobilesdkexample/walletsettings/WalletSettingsUtils.kt
package com.spruceid.mobilesdkexample.walletsettings

import com.spruceid.mobile.sdk.CredentialPack
import com.spruceid.mobile.sdk.KeyManager
import com.spruceid.mobile.sdk.rs.KeyStore
import com.spruceid.mobile.sdk.rs.Mdoc
import com.spruceid.mobile.sdk.rs.generateTestMdl
import com.spruceid.mobile.sdk.rs.generateTestPhotoId
import com.spruceid.mobilesdkexample.db.WalletActivityLogs
import com.spruceid.mobilesdkexample.utils.Toast
import com.spruceid.mobilesdkexample.utils.getCredentialIdTitleAndIssuer
import com.spruceid.mobilesdkexample.utils.getCurrentSqlDate
import com.spruceid.mobilesdkexample.viewmodels.CredentialPacksViewModel
import com.spruceid.mobilesdkexample.viewmodels.WalletActivityLogsViewModel

/** Signing key used as the DeviceKey of every locally generated test mdoc. */
const val TEST_MDOC_KEY_ALIAS = "testMdl"

suspend fun generateMockMdl(
    credentialPacksViewModel: CredentialPacksViewModel,
    walletActivityLogsViewModel: WalletActivityLogsViewModel
) = generateMockMdoc(
    displayName = "mDL",
    makeMdoc = ::generateTestMdl,
    credentialPacksViewModel = credentialPacksViewModel,
    walletActivityLogsViewModel = walletActivityLogsViewModel
)

suspend fun generateMockPhotoId(
    credentialPacksViewModel: CredentialPacksViewModel,
    walletActivityLogsViewModel: WalletActivityLogsViewModel
) = generateMockMdoc(
    displayName = "Photo ID",
    makeMdoc = ::generateTestPhotoId,
    credentialPacksViewModel = credentialPacksViewModel,
    walletActivityLogsViewModel = walletActivityLogsViewModel
)

/** Issue a test mdoc against the shared test signing key, store it, and log the activity. */
private suspend fun generateMockMdoc(
    displayName: String,
    makeMdoc: (KeyStore, String) -> Mdoc,
    credentialPacksViewModel: CredentialPacksViewModel,
    walletActivityLogsViewModel: WalletActivityLogsViewModel
) {
    try {
        val keyManager = KeyManager()
        if (!keyManager.keyExists(TEST_MDOC_KEY_ALIAS)) {
            keyManager.generateSigningKey(TEST_MDOC_KEY_ALIAS)
        }
        val mdoc = makeMdoc(keyManager, TEST_MDOC_KEY_ALIAS)
        val mdocPack = CredentialPack()

        val credentials = mdocPack.addMdoc(mdoc)
        credentialPacksViewModel.saveCredentialPack(mdocPack)

        val credentialInfo = getCredentialIdTitleAndIssuer(mdocPack, credentials[0])
        walletActivityLogsViewModel.saveWalletActivityLog(
            walletActivityLogs = WalletActivityLogs(
                credentialPackId = mdocPack.id().toString(),
                credentialId = credentialInfo.first,
                credentialTitle = credentialInfo.second,
                issuer = credentialInfo.third,
                action = "Claimed",
                dateTime = getCurrentSqlDate(),
                additionalInformation = ""
            )
        )

        Toast.showSuccess("Test $displayName added to your wallet")
    } catch (_: Exception) {
        Toast.showError("Error generating $displayName")
    }
}

File: ios/Showcase/Targets/AppUIKit/Sources/walletSettings/WalletSettingsUtils.swift
import Foundation
import SpruceIDMobileSdk
import SpruceIDMobileSdkRs

@MainActor
public func generateMockMdl() async {
    await generateMockMdoc(
        displayName: "mDL",
        makeMdoc: { keyManager, keyAlias in
            try generateTestMdl(keyManager: keyManager, keyAlias: keyAlias)
        }
    )
}

@MainActor
public func generateMockPhotoId() async {
    await generateMockMdoc(
        displayName: "Photo ID",
        makeMdoc: { keyManager, keyAlias in
            try generateTestPhotoId(keyManager: keyManager, keyAlias: keyAlias)
        }
    )
}

/// Issue a test mdoc against the default signing key, add it to the wallet, and log the activity.
@MainActor
private func generateMockMdoc(
    displayName: String,
    makeMdoc: (KeyManager, String) throws -> Mdoc
) async {
    do {
        ensureDefaultSigningKey()
        let mdoc = try makeMdoc(KeyManager(), DEFAULT_SIGNING_KEY_ID)
        let mdocPack = CredentialPack()

        let credentials = try await mdocPack.addMDoc(mdoc: mdoc)

        let bundle = Bundle.main
        let storageManager = StorageManager(
            appGroupId: bundle.object(forInfoDictionaryKey: "storageAppGroup") as? String)
        try await mdocPack.save(
            storageManager: storageManager
        )
        let credentialInfo = getCredentialIdTitleAndIssuer(
            credentialPack: mdocPack,
            credential: credentials[0]
        )
        _ = WalletActivityLogDataStore.shared.insert(
            credentialPackId: mdocPack.id.uuidString,
            credentialId: credentialInfo.0,
            credentialTitle: credentialInfo.1,
            issuer: credentialInfo.2,
            action: "Claimed",
            dateTime: Date(),
            additionalInformation: ""
        )

        ToastManager.shared.showSuccess(
            message: "Test \(displayName) added to your wallet"
        )

    } catch {
        print(error.localizedDescription)
        ToastManager.shared.showError(
            message: "Error generating \(displayName)"
        )
    }
}
